### PR TITLE
Use flake to pin darwin and nixpkgs inputs

### PR DESCRIPTION
**Pull request by cole-h**

### CODE REVIEW

File: configuration.nix [hunk 1]
```diff
@@ -32,8 +32,8 @@
     };
 
     nixPath = [
-      "nixpkgs=channel:nixpkgs-unstable"
-      "darwin=https://github.com/LnL7/nix-darwin/archive/master.tar.gz"
+      "nixpkgs=${inputs.nixpkgs}"
+      "darwin=${inputs.darwin}"
```
Comment (grahamc) [inline configuration.nix]: I wonder if we can outright delete them if we're going to use flakes to build the darwin config overall?

File: apply.sh [hunk 2]
```diff
@@ -49,20 +55,29 @@ fi
...
     set -eux
 fi
 
-export NIX_PATH=darwin-config="/nix/home/darwin-config/$CONFIG_TARGET":nixpkgs=channel:nixpkgs-unstable:darwin=https://github.com/LnL7/nix-darwin/archive/master.tar.gz
-
 sudo rm /etc/nix/nix.conf || true
 
-darwin-rebuild switch
+export NIX_PATH=darwin-config="$config/$CONFIG_TARGET":nixpkgs=$nixpkgs:darwin=$darwin
+darwin-rebuild switch --flake "$config"#"$(uname -m)"
```
Comment (grahamc) [inline apply.sh]: I wonder if we need to do the `switch` both here and above? Can we ~only do it here? We should probably have the `uname -m` bit be passed in as a CONFIG_... var for other use cases, too.

File: apply.sh [hunk 3]
```diff
@@ -49,20 +55,29 @@ fi
 cd /nix/home/darwin-config
 nix-shell -p git --run "git fetch $CONFIG_REPO $CONFIG_BRANCH && git checkout FETCH_HEAD"
 
+nix --extra-experimental-features 'nix-command flakes' profile install nixpkgs#git
+
+config="/nix/home/darwin-config"
+nixpkgs=$(nix --extra-experimental-features 'nix-command flakes' eval --raw "$config"#inputs.nixpkgs)
+darwin=$(nix --extra-experimental-features 'nix-command flakes' eval --raw "$config"#inputs.darwin)
+export NIX_PATH=darwin-config="$config/$CONFIG_TARGET":nixpkgs=$nixpkgs:darwin=$darwin
```
Comment (grahamc) [inline apply.sh]: I wonder if we could just ... delete these... :)

File: configuration.nix (hunk 1 above)
Comment (grahamc) [reply configuration.nix]: at any rate we probably don't want to pin nixpkgs at this level, so the buildkite environment is a bit more "floating."

File: apply.sh (hunk 3 above)
Comment (grahamc) [reply apply.sh]: (...and other refs to NIX_PATH here.)

**Review by grahamc — COMMENTED** [review]
initial post-con review, mostly just curious if we can prune additional cruft by virtue of what we're doing here. Looking pretty good overall!

File: apply.sh (hunk 2 above)
Comment (cole-h) [reply apply.sh]: We don't, I'm just mimicking what we had before -- the nix-darwin installer implicitly switches to a default config IIRC. But I guess that's just for bootstrapping the environment to be able to use nix-darwin properly, so I don't mind.

File: apply.sh (hunk 3 above)
Comment (cole-h) [reply apply.sh]: Maybe, I'll take a look.

Comment (cole-h) [reply apply.sh]: Yep, all done.

File: apply.sh [hunk 4]
```diff
@@ -5,6 +5,7 @@ set -eu
 CONFIG_REPO=$1
 CONFIG_BRANCH=$2
 CONFIG_TARGET=$3
+CONFIG_ARCH=$4 # arm64 or x86_64
```
Comment (cole-h) [inline apply.sh]: Uh, something something reimplementing flake URIs something.

File: apply.sh [hunk 5]
```diff
@@ -31,38 +35,19 @@ if ! hash nix; then
...
-export NIX_PATH=darwin-config="/nix/home/darwin-config/$CONFIG_TARGET":nixpkgs=channel:nixpkgs-unstable:darwin=https://github.com/LnL7/nix-darwin/archive/master.tar.gz
+# TODO(cole-h): maybe make the ref configurable as well? have to see if darwin-
+# rebuild accepts `darwinConfigurations.asdf` instead of just `asdf`
+nix --extra-experimental-features 'nix-command flakes' build "$CONFIG_FLAKE_URI"#darwinConfigurations."$CONFIG_ARCH".system
 
 sudo rm /etc/nix/nix.conf || true
 
-darwin-rebuild switch
+./result/sw/bin/darwin-rebuild switch --flake "$CONFIG_FLAKE_URI"#"$CONFIG_ARCH"
+unlink ./result
```
Comment (grahamc) [inline apply.sh]: It'd be lovely to call `nix build` once and call some program that comes out of that. Does nix-darwin's system output include an activate script, or can we pass it the result symlink as the profile, or something to that effect?

Also let's hold on to the result symlink, as a spare GC root.

I don't particularly like passing in the arch and the flake uri as two separate things.

**Review by grahamc — APPROVED** [review]

**Pull request merged**